CRM/Extension/Upgrader/SchemaTrait.php

```diff
@@ -72,4 +72,46 @@ public static function dropIndex($table, $indexName) {
     return TRUE;
   }
 
+  /**
+   * Create table (if not exists) from a given php schema file.
+   *
+   * The original entityType.php file should be copied and prefixed with the version-added.
+   *
+   * @param string $filePath
+   *   Absolute path to schema file (should be a copy not the original)
+   * @return bool
+   * @throws CRM_Core_Exception
+   */
+  public function createEntityTable(string $filePath): bool {
+    $entityDefn = include $filePath;
+    $schemaHelper = Civi::schemaHelper($this->getExtensionKey());
+    $sql = $schemaHelper->arrayToSql($entityDefn);
+    CRM_Core_DAO::executeQuery($sql, [], TRUE, NULL, FALSE, FALSE);
+    return TRUE;
+  }
+
+  /**
+   * Task to add or change a column definition, based on the php schema spec.
+   *
+   * @param string $entityName
+   * @param string $fieldName
+   * @param array $fieldSpec
+   *   As definied in the .entityType.php file for $entityName
+   * @return bool
+   * @throws CRM_Core_Exception
+   */
+  public function alterSchemaField(string $entityName, string $fieldName, array $fieldSpec): bool {
```

Comment:
@colemanw I know that this is the same as in core upgrader but any thought as to why we couldn't just read directly from the entitytype.php or is there some meta caching issue somewhere?

Reply:
@seamuslee001 because this is an upgrade step. You have to make a copy of the file because we need it to be frozen in time. Later, when the schema changes and the file is updated, it shouldn't inadvertently affect this upgrade step.